src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/Ordering/AbstractAddItemCommand.cs

```diff
@@ -67,19 +68,53 @@ protected override Task<CommandStatusResult> GetCommandStatusAsync(IProjectTree
         {
             var nodeToAddTo = GetNodeToAddTo(node);
 
-            // Call OnAddingNodesAsync.
-            // Then publish changes that could have taken place in OnAddingNodesAsync.
-            await OnAddingNodesAsync(nodeToAddTo).ConfigureAwait(true);
-            await _projectTree.TreeService.PublishLatestTreeAsync(waitForFileSystemUpdates: true).ConfigureAwait(true);
-
-            var updatedNode = _projectTree.CurrentTree.Find(node.Identity);
-            var updatedNodeToAddTo = _projectTree.CurrentTree.Find(nodeToAddTo.Identity);
-            var addedNodes = OrderingHelper.GetAddedNodes(nodeToAddTo, updatedNodeToAddTo);
+            var addedElements = await OrderingHelper.AddItems(_projectVsServices.ActiveConfiguredProject, _projectVsServices.ProjectLockService, () => OnAddingNodesAsync(nodeToAddTo)).ConfigureAwait(true);
 
-            if (addedNodes.Any())
+            if (addedElements.Any())
             {
-                // Make sure to change ConfigureAwait to "true" if we need switch back.
-                await OnAddedNodesAsync(_projectVsServices.ActiveConfiguredProject, node, addedNodes, updatedNode).ConfigureAwait(false);
+                await new ProjectAccessor(_projectVsServices.ProjectLockService).OpenProjectForWriteAsync(_projectVsServices.ActiveConfiguredProject, project => OnAddedElements(project, addedElements, node, nodeToAddTo)).ConfigureAwait(true);
+
+                // Re-select the node that was the target.
+                await _projectTree.TreeService.PublishLatestTreeAsync(waitForFileSystemUpdates: true).ConfigureAwait(true);
+                HACK_NodeHelper.Select(_projectVsServices.ActiveConfiguredProject, _serviceProvider, node);
```

Comment:
You need a switch to UI thread. Don't rely on being called on the UI thread, being async you are expected to switch to the thread you need.

Reply:
Addressed in the other PR.